pkg/resource/stack/secrets.go

```diff
@@ -125,132 +123,404 @@ func (csm *cachingSecretsManager) State() json.RawMessage {
+// This method is thread-safe and can be called concurrently by multiple goroutines.
+func (c *secretCache) Write(plaintext, ciphertext string, secret *resource.Secret) {
+	if c.disableCache {
+		return
 	}
-	ciphertext, err := csm.manager.Encrypter().EncryptValue(ctx, plaintext)
-	if err != nil {
-		return "", err
+	entry := secretCacheEntry{plaintext, ciphertext, secret}
+	c.bySecret.Store(secret, entry)
+	c.byCiphertext.Store(ciphertext, entry)
+}
+
+// TryEncrypt returns the cached ciphertext for the given secret and plaintext, if it exists.
+// The ciphertext is returned as a string, and a boolean is returned to indicate whether the secret was found.
+// This method is thread-safe and can be called concurrently by multiple goroutines.
+func (c *secretCache) TryEncrypt(secret *resource.Secret, plaintext string) (string, bool) {
+	entry, ok := c.bySecret.Load(secret)
+	if !ok {
+		return "", false
+	}
+	cacheEntry := entry.(secretCacheEntry)
+	if cacheEntry.plaintext != plaintext {
+		return "", false
 	}
-	csm.insert(secret, plaintext, ciphertext)
-	return ciphertext, nil
+	return cacheEntry.ciphertext, true
 }
 
-// insert associates the given secret with the given plain- and ciphertext in the cache.
-func (csm *cachingSecretsManager) insert(secret *resource.Secret, plaintext, ciphertext string) {
-	csm.cache[secret] = cacheEntry{plaintext, ciphertext}
+// TryDecrypt returns the cached plaintext for the given ciphertext, if it exists.
+// The plaintext is returned as a string, and a boolean is returned to indicate whether the ciphertext was found.
+// This method is thread-safe and can be called concurrently by multiple goroutines.
+func (c *secretCache) TryDecrypt(ciphertext string) (string, bool) {
+	entry, ok := c.byCiphertext.Load(ciphertext)
+	if !ok {
+		return "", false
+	}
+	return entry.(secretCacheEntry).plaintext, true
 }
 
-// mapDecrypter is a Decrypter with a preloaded cache. This decrypter is used specifically for deserialization,
-// where the deserializer is expected to prime the cache by scanning each resource for secrets, then decrypting all
-// of the discovered secrets en masse. Although each call to Decrypt _should_ hit the cache, a mapDecrypter does
-// carry an underlying Decrypter in the event that a secret was missed.
+// BatchEncrypter is a wrapper around an Encrypter that allows encryption operations to be batched to avoid round-trips
+// for when the underlying encrypter requires a network call. Encryptions to a secret are enqueued and processed in a
+// batch request either when the queue is full or when the wrapping transaction is committed.
+type BatchEncrypter interface {
+	config.Encrypter
+
+	// Enqueue adds a secret to the queue for encryption. If the cache has an entry for this secret and the plaintext has
+	// not changed, the previous ciphertext is used immediately. Otherwise, the secret is added to the queue.
+	// This method is thread-safe and can be called concurrently by multiple goroutines.
```

Comment:
Again, I'd pull caching out of the docs here -- it doesn't feel like a property of the interface to me.